src/transformers/models/patchtst/modeling_patchtst.py

```diff
@@ -0,0 +1,1958 @@
+        # partitioned across GPUs when using tensor-parallelism.
+        attn_output = attn_output.reshape(bsz, tgt_len, self.embed_dim)
+
+        attn_output = self.out_proj(attn_output)
+
+        return attn_output, attn_weights_reshaped, past_key_value
+
+
+class Transpose(nn.Module):
+    def __init__(self, *dims, contiguous=False):
+        super().__init__()
+        self.dims, self.contiguous = dims, contiguous
+
+    def forward(self, x):
+        if self.contiguous:
+            return x.transpose(*self.dims).contiguous()
+        else:
+            return x.transpose(*self.dims)
+
+
+def positional_encoding(pe, learn_pe, q_len, d_model):
+    # Positional encoding
+    if pe is None:
+        w_pos = torch.empty((q_len, d_model))  # pe = None and learn_pe = False can be used to measure impact of pe
+        nn.init.uniform_(w_pos, -0.02, 0.02)
+        learn_pe = False
+    elif pe == "zeros":
+        w_pos = torch.empty((q_len, d_model))
+        nn.init.uniform_(w_pos, -0.02, 0.02)
+    elif pe == "normal":
+        w_pos = torch.zeros((q_len, 1))
+        torch.nn.init.normal_(w_pos, mean=0.0, std=0.1)
+    elif pe == "uniform":
+        w_pos = torch.zeros((q_len, 1))
+        nn.init.uniform_(w_pos, a=0.0, b=0.1)
+    elif pe == "sincos":
+        pos_enc = torch.zeros(q_len, d_model)
+        position = torch.arange(0, q_len).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2) * -(math.log(10000.0) / d_model))
+        pos_enc[:, 0::2] = torch.sin(position * div_term)
+        pos_enc[:, 1::2] = torch.cos(position * div_term)
+        pos_enc = pos_enc - pos_enc.mean()
+        pos_enc = pos_enc / (pos_enc.std() * 10)
+        w_pos = pos_enc
+    else:
+        raise ValueError(
+            f"{pe} is not a valid positional encoder. Available types are 'normal', 'zeros', 'zero', uniform', 'sincos', None."
+        )
+    return nn.Parameter(w_pos, requires_grad=learn_pe)
+
+
+def random_masking(
+    xb: torch.Tensor,
+    mask_ratio: float,
+    unmasked_channel_indices: list = None,
+    channel_consistent_masking: bool = False,
+    mask_value=0,
+    seed_number: Optional[int] = None,
+):
+    """random_masking: Mask the input considering the control variables.
+
+    Parameters:
+        xb (Tensor): Input to mask [ bs x nvars x num_patches x patch_length]
+        mask_ratio (float): Mask ratio.
+        unmasked_channel_indices (list, optional):
+            indices of unmasked channels. These channels will not be masked. Defaults to None.
+        channel_consistent_masking (bool, optional):
+            When true, masking will be same across all channels of a timeseries. Otherwise, masking positions will vary
+            across channels. Defaults to True.
+        mask_value (int, optional): Value to use for masking. Defaults to 0.
+        seed_number (int, optional): Value to set for the random seed.
```

Comment:
Docstring here doesn't follow the HF format yet, see here for details: https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation

Reply:
Resolved